applications/AdjointFluidApplication/custom_utilities/response_function.h

```diff
@@ -0,0 +1,526 @@
+//  KratosAdjointFluidApplication
+//
+//  License:		 BSD License
+//					 license: AdjointFluidApplication/license.txt
+//
+//  Main authors:    Michael Andre, https://github.com/msandre
+//
+
+#if !defined(KRATOS_RESPONSE_FUNCTION)
+#define KRATOS_RESPONSE_FUNCTION
+
+// System includes
+#include <vector>
+#include <string>
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/element.h"
+#include "includes/process_info.h"
+#include "includes/model_part.h"
+#include "includes/kratos_parameters.h"
+#include "includes/ublas_interface.h"
+#include "utilities/openmp_utils.h"
+
+// Application includes
+
+namespace Kratos
+{
+///@addtogroup AdjointFluidApplication
+///@{
+
+///@name Kratos Classes
+///@{
+
+/// A base class for response functions.
+class ResponseFunction
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    KRATOS_CLASS_POINTER_DEFINITION(ResponseFunction);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Constructor.
+    ResponseFunction(ModelPart& rModelPart, Parameters& rParameters)
+      : mrModelPart(rModelPart)
+    {
+        KRATOS_TRY;
+
+        mSensitivityModelPartName =
+            rParameters["sensitivity_model_part_name"].GetString();
+        Parameters nodal_sensitivity_variables = rParameters["nodal_sensitivity_variables"];
+        mNodalSensitivityVariables.resize(nodal_sensitivity_variables.size());
+        for (unsigned int i = 0; i < nodal_sensitivity_variables.size(); ++i)
+            mNodalSensitivityVariables[i] = nodal_sensitivity_variables[i].GetString();
+
+        KRATOS_CATCH("");
+    }
+
+    /// Destructor.
+    virtual ~ResponseFunction()
+    {
+    }
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    ModelPart& GetModelPart()
+    {
+      return mrModelPart;
+    }
+
+    const ModelPart& GetModelPart() const
+    {
+      return mrModelPart;
+    }
+
+    virtual void Initialize()
+    {
+        KRATOS_TRY;
+
+        this->Clear();
+        this->Check();
+
+        ModelPart& r_model_part = this->GetModelPart();
+
+        // Initialize flags.
+#pragma omp parallel
```

Comment:
we need to find a better way to have this parallel loops. I'll work on it. fine for now...